### DISCUSSION TITLE
toolchains clang script no working

### DISCUSSION THREAD

**Discussion by Tracker647** [android/ndk#2050] [Q&A] [answered]
I'm trying cross compiling cairo 1.18.0 in android, which is use meson build system.
my environment:
Terminal: Windows 11 WSL Ubuntu 22.04
NDK: linux-r25c

cross compile file is:
```
[constants]
ndk_path = '/mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c'
toolchain = ndk_path / 'toolchains/llvm/prebuilt/linux-x86_64'

[host_machine]
system = 'android'
cpu_family = 'x86'
cpu = 'i686'
endian = 'little'

[properties]
sys_root = ndk_path / 'sysroot'
pkg_config_libdir = '/usr/bin/'
c_ld = 'gold'
cpp_ld = 'gold'

[binaries]
c = toolchain / 'bin/i686-linux-android23-clang'
cpp = toolchain / 'bin/i686-linux-android23-clang++'
ar = toolchain / 'bin/llvm-ar'
strip = toolchain / 'bin/llvm-strip'

```

when I execute meson the error occurs:
```
root@WINDOWS-1FNAO0N:/mnt/d/CodeDraft/android_ndk_test/3rd/cairo-1.18.0# meson setup --cross-file cross_file_x86.txt build_x86
The Meson build system
Version: 1.5.0
Source dir: /mnt/d/CodeDraft/android_ndk_test/3rd/cairo-1.18.0
Build dir: /mnt/d/CodeDraft/android_ndk_test/3rd/cairo-1.18.0/build_x86
Build type: cross build
Program version.py found: YES (/mnt/d/CodeDraft/android_ndk_test/3rd/cairo-1.18.0/version.py)
Project name: cairo
Project version: 1.18.0

meson.build:1:0: ERROR: Unknown compiler(s): [['/mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android23-clang']]

A full log can be found at /mnt/d/CodeDraft/android_ndk_test/3rd/cairo-1.18.0/build_x86/meson-logs/meson-log.txt
```
check log, seems the clang script is not received arguments properly:
```
Project name: cairo
Project version: 1.18.0
-----------
Detecting compiler via: `/mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android23-clang --version` -> 1
stderr:
clang: error: no input files
-----------
```

`cd` to ndk path and executive, it keeps:
```
root@WINDOWS-1FNAO0N:/mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin# ./i686-linux-android23-clang -v
clang: error: no input files
```

and I found clang and clang++ files in r25c bin just simple txt
clang
```
clang-14
```
clang++
```
clang
```

execute clang-14 in bin, it works:
```
root@WINDOWS-1FNAO0N:/mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin# ./clang-14 -v
Android (9352603, based on r450784d1) clang version 14.0.7 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /mnt/d/Devtools/android-ndk-r25c-linux/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/.
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Candidate multilib: .;@m64
Selected multilib: .;@m64
```

I have another r26d version, it works normal, the difference of bin directory with r25c is that the clang and clang++ is actual executable binary, no txt, it download about 3 months ago, now I found the same r26d version downloaded from the NDK official site have clang and clang++ are txt files just like r25c and its later version r27, and same compile error in meson.

I have no idea whether is any configure I need be done or I just choose the false download repository?

**Comment by Tracker647** [comment]
ok, in my r26d the clang and clang++ is not executable file but symlinks, but I still don't know why my r26d is different from the same version I downloaded from official site since then.
```
lrwxrwxrwx 1 root root         8 Jul  9 10:50 clang -> clang-17*
lrwxrwxrwx 1 root root         5 Jul  9 10:50 clang++ -> clang*
```

**Comment by DanAlbert** [comment, marked as answer]
Can you migrate to not using those? There's really no reason to use them. You can replace `c = toolchain / 'bin/i686-linux-android23-clang'` with `c = toolchain / 'bin/clang --target=i686-linux-android23'. It's the same intended effect without the occasional bugs introduced by the intermediate script. The only reason the trampolines exist is to work in build systems that for some reason can't handle using clang normally.

I see that https://developer.android.com/ndk/guides/other_build_systems recommends both, but then the examples use only the thing I just said to not do. I've just sent a CL to make it clear on that page that `--target` should be preferred, and the prefixed launchers are a last resort.

If you want them fixed, file a bug with a repro case (though it seems like this may just be an install failure, and theres's nothing to fix on our end).

> and I found clang and clang++ files in r25c bin just simple txt

This is what happens when a zip with symlinks in it is extracted by a tool that doesn't know how to handle those. You probably used different tools to extract each NDK, and not all of them were capable of understanding symlinks.

**Reply by Tracker647** [reply]
the txt problem solved, thanks.
base your zip tool problem hint I try windows self-contained decompress tool and bandizip, found if I unzip the file with windows decompresser the clang and clang++ become txt files, and it become normal symlink if use bandizip with adminstrator.
plus, the built-in unzip command in wsl also works.

**Reply by DanAlbert** [reply]
Oh, I'd missed that this was a WSL thing. Yeah, Windows doesn't do symlinks without admin, so the Windows NDK doesn't contain any symlinks. If you extract the Linux NDK on Windows, you need to use an archiver that can handle those.
